media/js/firefox/quantum/features-scroller.js

```diff
@@ -0,0 +1,216 @@
+                video.pause();
+            }
+        } catch(e) {
+            // Fail silently.
+        }
+    }
+
+    function scrollToPrevFeature() {
+        var currentElem = document.querySelector('.features-scroller-nav a.current');
+
+        if (currentElem) {
+            var prevElem = currentElem.parentNode.previousElementSibling;
+
+            if (prevElem) {
+                prevElem.firstChild.click();
+            }
+        }
+    }
+
+    function scrollToNextFeature() {
+        var currentElem = document.querySelector('.features-scroller-nav a.current');
+
+        if (currentElem) {
+            var nextElem = currentElem.parentNode.nextElementSibling;
+
+            if (nextElem) {
+                nextElem.firstChild.click();
+            }
+        }
+    }
+
+    function initNavigation() {
+        var navLinks = document.querySelectorAll('.features-scroller-nav a');
+        var prevButton = document.querySelector('.features-scroller-nav .previous');
+        var nextButton = document.querySelector('.features-scroller-nav .next');
+
+        for (var i = 0; i < navLinks.length; i++) {
+            navLinks[i].addEventListener('click', scrollToFeature, false);
+        }
+
+        prevButton.addEventListener('click', scrollToPrevFeature, false);
+        nextButton.addEventListener('click', scrollToNextFeature, false);
+    }
+
+    function trackGAScroll(el) {
+        var label = el.getAttribute('data-ga-label');
+
+        if (label) {
+            // remove attributes so we only track scroll for each element once.
+            el.removeAttribute('data-ga-label');
+
+            window.dataLayer.push({
+                'eAction': 'scroll',
+                'eLabel': label,
+                'event': 'non-interaction'
+            });
+        }
+    }
+
+    function initWaypoints() {
+        var features = document.querySelectorAll('.features-scroller-content > .feature-content');
+
+        for (var i = 0; i < features.length; i++) {
+            featureWaypoints.push(new Waypoint({
+                element: features[i],
+                handler: function(direction) {
+                    if (direction === 'down') {
+                        setActiveFeature(this.element.id);
+                        playVideo(this.element);
+                        trackGAScroll(this.element);
+                    } else {
+                        pauseVideo(this.element);
+                    }
+                },
+                offset: '50%'
+            }));
+
+            featureWaypoints.push(new Waypoint({
+                element: features[i],
+                handler: function(direction) {
+                    if (direction === 'up') {
+                        setActiveFeature(this.element.id);
+                        playVideo(this.element);
+                    } else {
+                        pauseVideo(this.element);
+                    }
+                },
+                offset: '-50%'
+            }));
```

Comment:
Not sure I see a way around it (aside from not auto-playing!), but having two waypoints per feature is a bummer.

Reply:
Yeah I thought about a way to do this too, but if we're to support auto play for both scrolling up and down the web page, two waypoints are needed: one for when the videos pass through the top of the viewport, and one for when the pass through the bottom of the viewport. This seems to work well enough, even though it's more overhead than i would have liked. In a perfect world, I would have not chosen to auto-play the videos at all (I think it's arrogant to assume users want this).